### ISSUE TITLE
[compiler] `export { symbol as ɵ }` duplicates `export { symbol }

### ISSUE THREAD

**Issue by dherges** [angular/angular#21058]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report 
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

When using `flatModuleOutFile`, the generated bundle index and the metadata json may export symbols twice.

Example for bundle index:

```ts
/* (in bundle index) */
export { ControlAndPluginService as ɵb , ... }

/* (referenced through bundle index -> ./public_api.d.ts -> typescript barrel -> ./index.d.ts) */
export { ControlAndPluginService }
```

Summary and repro here:
https://github.com/ghiscoding/Angular-Slickgrid/commit/dc8447d8aa503dace199587a5cfa31bbffae2595#r26315557

Example for AoT metadata:

```json
  "origins": {
    "ControlAndPluginService": ".\/src\/app\/modules\/angular-slickgrid\/services\/controlAndPlugin.service",
    "\u0275a": ".\/src\/app\/modules\/angular-slickgrid\/services\/odata.service",
    "\u0275b": ".\/src\/app\/modules\/angular-slickgrid\/services",
    "\u0275c": ".\/src\/app\/modules\/angular-slickgrid\/services",
    "\u0275d": ".\/src\/app\/modules\/angular-slickgrid\/services",
    "\u0275e": ".\/src\/app\/modules\/angular-slickgrid\/services",
    "\u0275f": ".\/src\/app\/modules\/angular-slickgrid\/services",
    "\u0275g": ".\/src\/app\/modules\/angular-slickgrid\/services"
  }
```

Repro here: https://github.com/ghiscoding/Angular-Slickgrid/blob/eaa7f3a0d55748938ca4211bf1f561661d9cdb98/dist/angular-slickgrid.metadata.json

```json
 "origins": {
    "SelectableListModule": "./selectable-list-module",
    "\u0275a": "./directive\/selectable-list\/selectable-list.directive",
    "SelectableListDirective": "./directive\/selectable-list\/selectable-list.directive"
  }
```

Repro here: https://github.com/darkbasic/ngx-selectable-list/tree/packagr

## Expected behavior

tbd (shouldn't the symbols be exported just once?)

## Minimal reproduction of the problem with instructions

dherges/ng-packagr#377
dherges/ng-packagr#382

## What is the motivation / use case for changing the behavior?

AoT compatibility. After distributing erroneous packages, the duplicated exports seem to fail an app's build.

## Environment

<pre><code>
Angular version: X.Y.Z 
For Tooling issues:
- Node version: XX  
- Platform:  

Others:
</code></pre>

---

// fyi @hansl @filipesilva 

Are these the issues you mentioned at AngularConnect? I eventually received two repros that helped to discover it.

**Comment by filipesilva** [comment]
@dherges I wasn't aware of this particular issue no. But that sure would explain why some libraries fail with no discernible reason...

**Comment by chuckjaz** [comment]
@dherges The `"origins"` list is not the list of exports. It is a mapping from the symbol to the module that it was originally declared in. Duplications are expected.

The second example you provided can be fixed by making the following change to `selectable-list.directive.ts` in `ngx-selectable-list`.

```
--- a/src/app/selectable-list/directive/selectable-list/selectable-list.directive.ts
+++ b/src/app/selectable-list/directive/selectable-list/selectable-list.directive.ts
@@ -21,12 +21,14 @@ export enum Mode {
   both = 'Single and multiple (press activated)',
 }
 
+export const SERVICE = new SelectableListService();
+
 @Directive({
   selector: '[appSelectableList]',
   providers: [
     {
       provide: 'selectable-list-service',
-      useValue: new SelectableListService(),
+      useValue: SERVICE,
     }
   ],
 })
```

This rewriting is performed automatically when the component is used directly in an application. However, the metadata that is bundled reflects the metadata before this rewriting.

I reported this issue as #21082.

I could not decipher  https://github.com/dherges/ng-packagr/issues/377.  Could you determine this is a similar metadata error in disguise?

**Comment by dherges** [comment]
@chuckjaz From your explanation, the example in dherges/ng-packagr#377 is then a valid metadata JSON. I noticed the duplicated exports in the `*.d.ts`, too. Duplicated exports are not the issue I guess?!

Adding @ghiscoding to the issue discussion: can you clarify? You reported the error in an application's prod build:

```
ERROR in Error: Can't resolve all parameters for AngularSlickgridComponent in C:/sourcecode/Angular-Slickgrid-PackageTest/node_modules/angular-slickgrid/angular-slickgrid.d.ts: (?, ?, ?, ?, ?, ?).
    at syntaxError (C:\sourcecode\Angular-Slickgrid-PackageTest\node_modules\@angular\compiler\bundles\compiler.umd.js:1729:34)
```

The AoT metadata for the component includes references to "greek-b" through to "greek-g":

```json
"__ctor__": [
          {
            "__symbolic": "constructor",
            "parameters": [
              {
                "__symbolic": "reference",
                "name": "\u0275b"
              },
              {
                "__symbolic": "reference",
                "name": "\u0275c"
              },
              {
                "__symbolic": "reference",
                "name": "\u0275d"
              },
              {
                "__symbolic": "reference",
                "name": "\u0275e"
              },
              {
                "__symbolic": "reference",
                "name": "\u0275f"
              },
              {
                "__symbolic": "reference",
                "name": "\u0275g"
              }
            ]
          }
        ]
```

I just now noticed that the "greek-b" to "greek-g" symbols are not listed in the `exports` of the metadata JSON. It is exported in the ESM bundle index ([angular-slickgrid.js](https://raw.githubusercontent.com/ghiscoding/Angular-Slickgrid/2ce40ce0fba9963f9c82c68acb89599e1b362419/dist/angular-slickgrid.js)):

```js
export  { .. ControlAndPluginService as ɵb .. }
```

**Comment by ghiscoding** [comment]
@dherges I'm actually not sure why and where these symbolic greek symbols comes from. I thought it was done by your packager? Anyhow, it doesn't come directly from my code directly. 

As for the exports and the unresolved dependencies, I previously had the following
1. an `index.ts` inside my Services folder that included all exports 
2. I also re-exported all these services them through the `public_api.ts` for external use
3. Inside the lib component, I was using the imports (for the DI) with `import { serviceA, serviceB, ... } from './services';`. (from step 1).

My refactor was
4. Delete the Services `index.ts` (that was step1) 
5. Import each Services 1 by 1, `import { serviceA } from './services/serviceA';`

Does that answer your questions?

**Comment by chuckjaz** [comment]
@ghiscoding The latin-barred-o (`ɵ`) comes from the angular bundler as part of `ngc` used by `packagr`.

`packagr` uses `ngc` to produce FESM which is a flat module. The flattened version of the metadata is produced by `ngc` using the `flatModuleOutFile` and `flatModuleId` options passed to `ngc` via a tsconfig.json produced by `packagr` based on the `packagr` metadata supplied to it.

What `ngc` does is traverse all the symbols exported from root FESM module. It then extracts the metadata and creates a "shrouded public" (that is, a latin-barred-o prefix identifier) for each symbol that is referenced by metadata but is not publicly exported by the root FESM module. This allows the compiler to generate an import using the `flatModuleId` without having to use a deep import that would not work in many bundlers.

In other words,  a `ɵ` prefixed identifier indicates a metadata reference exists that is needed by AOT for  and that identifier and it was not exported by the root index of the FESM.

if these exist in the constructor member then it is an indication that the types used in the constructor of a type where not exported from the FESM root module.

**Comment by ghiscoding** [comment]
Thanks @chuckjaz for the explanation, however I must say that I very recently (less than 2 months ago) switched from AngularJS to Angular 4+ and all of these new terms (FESM, rollup, AOT, JIT, tree shacking, ...) is rather a lot for me to digest. However my work needed a datagrid library that didn't seem to be available with all our need (we previously used UI-Grid) and so I took the task of porting a jQuery datagrid lib to Angular 4 and this is where it all started, quite a task for someone new to the Angular 4+ world.

Anyhow, if you got more questions, then let me know and thanks for the help and informative subject :)

**Comment by chuckjaz** [comment]
@dherges I am closing this issue as there doesn't seem to be anything for me to fix here. Please feel free to open another issue if there is something I missed.

**Comment by dherges** [comment]
Hi @chuckjaz, thanks for the good explanation of "shrouded publics"! Is it then valid that shrouded publics are referred in the metadata but not exported? My understanding used to be that shrouded publics must be exported so that the Angular compiler (application build) "picks up" the symbols.

**Comment by chuckjaz** [comment]
@dherges:
> Is it then valid that shrouded publics are referred in the metadata but not exported? 

No. Do you have an instance of that?

**Comment by dherges** [comment]
Here: https://github.com/ghiscoding/Angular-Slickgrid/blob/eaa7f3a0d55748938ca4211bf1f561661d9cdb98/dist/angular-slickgrid.metadata.json

```json
{
  "__symbolic": "module",
  "version": 3,
  "exports": [
    {
      "export": [
        {
          "name": "Editors",
          "as": "Editors"
        }
      ],
      "from": ".\/src\/app\/modules\/angular-slickgrid\/editors"
    },
    {
      "export": [
        {
          "name": "FilterConditions",
          "as": "FilterConditions"
        }
      ],
      "from": ".\/src\/app\/modules\/angular-slickgrid\/filter-conditions"
    },
    {
      "export": [
        {
          "name": "FilterTemplates",
          "as": "FilterTemplates"
        }
      ],
      "from": ".\/src\/app\/modules\/angular-slickgrid\/filter-templates"
    },
    {
      "export": [
        {
          "name": "Formatters",
          "as": "Formatters"
        }
      ],
      "from": ".\/src\/app\/modules\/angular-slickgrid\/formatters"
    },
    {
      "export": [
        {
          "name": "Sorters",
          "as": "Sorters"
        }
      ],
      "from": ".\/src\/app\/modules\/angular-slickgrid\/sorters"
    }
  ],
  "metadata": {
    /* ... (omitted for brevity) */
    "AngularSlickgridComponent": {
      "__symbolic": "class",
      "decorators": [/* ... */]
      "members": {
         /* ... */
        "__ctor__": [
          {
            "__symbolic": "constructor",
            "parameters": [
              {
                "__symbolic": "reference",
                "name": "\u0275b"
              },
              {
                "__symbolic": "reference",
                "name": "\u0275c"
              },
              {
                "__symbolic": "reference",
                "name": "\u0275d"
              },
              {
                "__symbolic": "reference",
                "name": "\u0275e"
              },
              {
                "__symbolic": "reference",
                "name": "\u0275f"
              },
              {
                "__symbolic": "reference",
                "name": "\u0275g"
              }
            ]
          }
    "\u0275a": {
      "__symbolic": "class",
      /* ... */
    }
    /*** --> \u275b, \u275c, and so on missing? ***/
  }
```

**Comment by dherges** [comment]
`ɵb` to `ɵg` are exported in the JavaScript bundle index. My question is: should shrouded publics be in `metadata` and/or `exports` of the AoT metadata JSON?

```js
/**
 * Generated bundle index. Do not edit.
 */

export { /* ... */ AngularSlickgridComponent, AngularSlickgridModule, ControlAndPluginService as ɵb, FilterService as ɵe, GridEventService as ɵc, GridExtraService as ɵd, ResizerService as ɵf, SortService as ɵg, OdataService as ɵa };
```

**Comment by chuckjaz** [comment]
@dherges:

> should shrouded publics be in metadata and/or exports of the AoT metadata JSON?

No. The exports information is a duplicate of what is in the original `.ts` file. It is included here so we don't have to both parse the `.metadata.json` file and the `.d.ts` file when analyzing metadata which save us a bit of time.

**Comment by ivanwonder** [comment]
I encounter the problem. I found NGC regard my exported `primary.component` as private. This is an example. Try to package the library `apf`.
https://github.com/ivanwonder/ng-packagr/tree/master/integration/samples/apf
Here, the `resolveModule` returns a wrong result.
https://github.com/angular/angular/blob/388c1a36384ea0c2abfd6b0509de6fd52936c699/packages/compiler-cli/src/metadata/bundler.ts#L185
When in the module `./src`, the `exportDeclaration.from` is `./primary.component`, the return path is `./primary.component`, the right should be `./src/primary.component`. I think the path `./src` shoule be resolved to `./src/index`.
@chuckjaz I hope this is helpful.
